Client 3.8.4+ with near cache gets java.lang.UnsupportedOperationException when connected to 3.7.x cluster

When Java client 3.8.4+ is configured with near cache and connects to a 3.7.x cluster, it continously gets the following exception:

```
Caused by: java.lang.UnsupportedOperationException: Unrecognized client message received with type: 0x142
    at com.hazelcast.client.impl.protocol.task.NoSuchMessageTask.processMessage(NoSuchMessageTask.java:50)
    at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.initializeAndProcessMessage(AbstractMessageTask.java:123)
    at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.run(AbstractMessageTask.java:103)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
    at java.lang.Thread.run(Thread.java:745)
    at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76)
    at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:92)
    at ------ submitted from ------.(Unknown Source)
    at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:95)
    ... 17 more
```

The operations are working as expected, but this is exception is thrown continuously.

--- a/hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/AbstractClientInternalCacheProxy.java
+++ b/hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/AbstractClientInternalCacheProxy.java
@@ -19,12 +19,9 @@ package com.hazelcast.client.cache.impl;
 import com.hazelcast.cache.impl.CacheEventData;
 import com.hazelcast.cache.impl.CacheEventListenerAdaptor;
 import com.hazelcast.cache.impl.CacheProxyUtil;
-import com.hazelcast.cache.impl.CacheService;
 import com.hazelcast.cache.impl.CacheSyncListenerCompleter;
 import com.hazelcast.cache.impl.operation.MutableOperation;
 import com.hazelcast.client.config.ClientConfig;
-import com.hazelcast.client.connection.ClientConnectionManager;
-import com.hazelcast.client.connection.nio.ClientConnection;
 import com.hazelcast.client.impl.ClientMessageDecoder;
 import com.hazelcast.client.impl.HazelcastClientInstanceImpl;
 import com.hazelcast.client.impl.protocol.ClientMessage;
@@ -41,8 +38,6 @@ import com.hazelcast.client.impl.protocol.codec.CacheRemoveAllKeysCodec;
 import com.hazelcast.client.impl.protocol.codec.CacheRemoveCodec;
 import com.hazelcast.client.impl.protocol.codec.CacheRemoveEntryListenerCodec;
 import com.hazelcast.client.impl.protocol.codec.CacheReplaceCodec;
-import com.hazelcast.client.spi.ClientClusterService;
-import com.hazelcast.client.spi.ClientContext;
 import com.hazelcast.client.spi.ClientListenerService;
 import com.hazelcast.client.spi.EventHandler;
 import com.hazelcast.client.spi.impl.ClientInvocation;
@@ -56,15 +51,13 @@ import com.hazelcast.config.NearCacheConfig;
 import com.hazelcast.core.ExecutionCallback;
 import com.hazelcast.core.HazelcastInstanceNotActiveException;
 import com.hazelcast.core.ICompletableFuture;
-import com.hazelcast.nio.IOUtil;
 import com.hazelcast.internal.adapter.ICacheDataStructureAdapter;
 import com.hazelcast.internal.nearcache.NearCache;
 import com.hazelcast.internal.nearcache.NearCacheManager;
 import com.hazelcast.internal.nearcache.impl.invalidation.RepairingHandler;
 import com.hazelcast.internal.nearcache.impl.invalidation.RepairingTask;
 import com.hazelcast.logging.ILogger;
-import com.hazelcast.nio.Address;
-import com.hazelcast.nio.Connection;
+import com.hazelcast.nio.IOUtil;
 import com.hazelcast.nio.serialization.Data;
 import com.hazelcast.spi.serialization.SerializationService;
 import com.hazelcast.util.ExceptionUtil;
@@ -93,7 +86,6 @@ import static com.hazelcast.cache.impl.operation.MutableOperation.IGNORE_COMPLET
 import static com.hazelcast.config.InMemoryFormat.NATIVE;
 import static com.hazelcast.config.NearCacheConfig.LocalUpdatePolicy.CACHE;
 import static com.hazelcast.config.NearCacheConfig.LocalUpdatePolicy.CACHE_ON_UPDATE;
-import static com.hazelcast.instance.BuildInfo.UNKNOWN_HAZELCAST_VERSION;
 import static com.hazelcast.instance.BuildInfo.calculateVersion;
 import static com.hazelcast.internal.nearcache.NearCacheRecord.NOT_RESERVED;
 import static com.hazelcast.util.ExceptionUtil.rethrow;
@@ -983,14 +975,19 @@ abstract class AbstractClientInternalCacheProxy<K, V> extends AbstractClientCach
 
         @Override
         public void beforeListenerRegister() {
-            pre38EventHandler.beforeListenerRegister();
             repairingEventHandler.beforeListenerRegister();
+
+            supportsRepairableNearCache = supportsRepairableNearCache();
+
+            if (!supportsRepairableNearCache) {
+                pre38EventHandler.beforeListenerRegister();
+
+                logger.warning(format("Near Cache for '%s' cache is started in legacy mode", name));
+            }
         }
 
         @Override
         public void onListenerRegister() {
-            supportsRepairableNearCache = supportsRepairableNearCache();
-
             if (supportsRepairableNearCache) {
                 repairingEventHandler.onListenerRegister();
             } else {
@@ -1015,16 +1012,17 @@ abstract class AbstractClientInternalCacheProxy<K, V> extends AbstractClientCach
     private final class RepairableNearCacheEventHandler extends CacheAddNearCacheInvalidationListenerCodec.AbstractEventHandler
             implements EventHandler<ClientMessage> {
 
-        private final RepairingHandler repairingHandler;
-
-        public RepairableNearCacheEventHandler() {
-            getRepairingTask().deregisterHandler(nameWithPrefix);
-            repairingHandler = getRepairingTask().registerAndGetHandler(nameWithPrefix, nearCache);
-        }
+        private volatile RepairingHandler repairingHandler;
 
         @Override
         public void beforeListenerRegister() {
-            // NOP
+            if (supportsRepairableNearCache()) {
+                RepairingTask repairingTask = getContext().getRepairingTask(getServiceName());
+                repairingHandler = repairingTask.registerAndGetHandler(nameWithPrefix, nearCache);
+            } else {
+                RepairingTask repairingTask = getContext().getRepairingTask(getServiceName());
+                repairingTask.deregisterHandler(nameWithPrefix);
+            }
         }
 
         @Override
@@ -1042,16 +1040,11 @@ abstract class AbstractClientInternalCacheProxy<K, V> extends AbstractClientCach
                            Collection<UUID> partitionUuids, Collection<Long> sequences) {
             repairingHandler.handle(keys, sourceUuids, partitionUuids, sequences);
         }
-
-        private RepairingTask getRepairingTask() {
-            ClientContext clientContext = getContext();
-            return clientContext.getRepairingTask(CacheService.SERVICE_NAME);
-        }
     }
 
     /**
      * This event handler is here to be used with server versions < 3.8.
-     * <p>
+     *
      * If server version is < 3.8 and client version is >= 3.8, this event handler must be used to
      * listen near cache invalidations. Because new improvements for near cache eventual consistency cannot work
      * with server versions < 3.8.
@@ -1156,23 +1149,6 @@ abstract class AbstractClientInternalCacheProxy<K, V> extends AbstractClientCach
                 return CacheRemoveEntryListenerCodec.decodeResponse(clientMessage).response;
             }
         };
-    }
-
-    private int getConnectedServerVersion() {
-        ClientContext clientContext = getContext();
-        ClientClusterService clusterService = clientContext.getClusterService();
-        Address ownerConnectionAddress = clusterService.getOwnerConnectionAddress();
-
-        HazelcastClientInstanceImpl client = getClient();
-        ClientConnectionManager connectionManager = client.getConnectionManager();
-        Connection connection = connectionManager.getConnection(ownerConnectionAddress);
-        if (connection == null) {
-            logger.warning(format("No owner connection is available, "
-                    + "near cached cache %s will be started in legacy mode", name));
-            return UNKNOWN_HAZELCAST_VERSION;
-        }
-
-        return ((ClientConnection) connection).getConnectedServerVersion();
     }
 
     private boolean supportsRepairableNearCache() {

--- a/hazelcast-client/src/main/java/com/hazelcast/client/proxy/NearCachedClientMapProxy.java
+++ b/hazelcast-client/src/main/java/com/hazelcast/client/proxy/NearCachedClientMapProxy.java
@@ -16,16 +16,12 @@
 
 package com.hazelcast.client.proxy;
 
-import com.hazelcast.client.connection.ClientConnectionManager;
-import com.hazelcast.client.connection.nio.ClientConnection;
-import com.hazelcast.client.impl.HazelcastClientInstanceImpl;
 import com.hazelcast.client.impl.protocol.ClientMessage;
 import com.hazelcast.client.impl.protocol.codec.MapAddNearCacheEntryListenerCodec;
 import com.hazelcast.client.impl.protocol.codec.MapAddNearCacheInvalidationListenerCodec;
 import com.hazelcast.client.impl.protocol.codec.MapGetAllCodec;
 import com.hazelcast.client.impl.protocol.codec.MapRemoveCodec;
 import com.hazelcast.client.impl.protocol.codec.MapRemoveEntryListenerCodec;
-import com.hazelcast.client.spi.ClientClusterService;
 import com.hazelcast.client.spi.ClientContext;
 import com.hazelcast.client.spi.EventHandler;
 import com.hazelcast.client.spi.impl.ListenerMessageCodec;
@@ -43,8 +39,6 @@ import com.hazelcast.map.EntryProcessor;
 import com.hazelcast.monitor.LocalMapStats;
 import com.hazelcast.monitor.NearCacheStats;
 import com.hazelcast.monitor.impl.LocalMapStatsImpl;
-import com.hazelcast.nio.Address;
-import com.hazelcast.nio.Connection;
 import com.hazelcast.nio.serialization.Data;
 import com.hazelcast.query.Predicate;
 import com.hazelcast.util.CollectionUtil;
@@ -60,12 +54,10 @@ import java.util.UUID;
 import java.util.concurrent.TimeUnit;
 
 import static com.hazelcast.core.EntryEventType.INVALIDATION;
-import static com.hazelcast.instance.BuildInfo.UNKNOWN_HAZELCAST_VERSION;
 import static com.hazelcast.instance.BuildInfo.calculateVersion;
 import static com.hazelcast.internal.nearcache.NearCache.CACHED_AS_NULL;
 import static com.hazelcast.internal.nearcache.NearCache.NOT_CACHED;
 import static com.hazelcast.internal.nearcache.NearCacheRecord.NOT_RESERVED;
-import static com.hazelcast.map.impl.MapService.SERVICE_NAME;
 import static com.hazelcast.util.ExceptionUtil.rethrow;
 import static java.lang.String.format;
 import static java.util.Collections.emptyMap;
@@ -529,7 +521,7 @@ public class NearCachedClientMapProxy<K, V> extends ClientMapProxy<K, V> {
             return;
         }
 
-        getContext().getRepairingTask(SERVICE_NAME).deregisterHandler(name);
+        getContext().getRepairingTask(getServiceName()).deregisterHandler(name);
         deregisterListener(invalidationListenerId);
     }
 
@@ -540,20 +532,26 @@ public class NearCachedClientMapProxy<K, V> extends ClientMapProxy<K, V> {
      */
     private final class ConnectedServerVersionAwareNearCacheEventHandler implements EventHandler<ClientMessage> {
 
-        private volatile boolean supportsRepairableNearCache;
-        private final RepairableNearCacheEventHandler repairingEventHandler = new RepairableNearCacheEventHandler();
         private final Pre38NearCacheEventHandler pre38EventHandler = new Pre38NearCacheEventHandler();
+        private final RepairableNearCacheEventHandler repairingEventHandler = new RepairableNearCacheEventHandler();
+
+        private volatile boolean supportsRepairableNearCache;
 
         @Override
         public void beforeListenerRegister() {
             repairingEventHandler.beforeListenerRegister();
-            pre38EventHandler.beforeListenerRegister();
+
+            supportsRepairableNearCache = supportsRepairableNearCache();
+
+            if (!supportsRepairableNearCache) {
+                pre38EventHandler.beforeListenerRegister();
+
+                logger.warning(format("Near Cache for '%s' map is started in legacy mode", name));
+            }
         }
 
         @Override
         public void onListenerRegister() {
-            supportsRepairableNearCache = supportsRepairableNearCache();
-
             if (supportsRepairableNearCache) {
                 repairingEventHandler.onListenerRegister();
             } else {
@@ -578,17 +576,17 @@ public class NearCachedClientMapProxy<K, V> extends ClientMapProxy<K, V> {
     private final class RepairableNearCacheEventHandler extends MapAddNearCacheInvalidationListenerCodec.AbstractEventHandler
             implements EventHandler<ClientMessage> {
 
-        private final RepairingHandler repairingHandler;
-
-        public RepairableNearCacheEventHandler() {
-            RepairingTask repairingTask = getContext().getRepairingTask(SERVICE_NAME);
-            repairingTask.deregisterHandler(name);
-            repairingHandler = repairingTask.registerAndGetHandler(name, nearCache);
-        }
+        private volatile RepairingHandler repairingHandler;
 
         @Override
         public void beforeListenerRegister() {
-            // NOP
+            if (supportsRepairableNearCache()) {
+                RepairingTask repairingTask = getContext().getRepairingTask(getServiceName());
+                repairingHandler = repairingTask.registerAndGetHandler(name, nearCache);
+            } else {
+                RepairingTask repairingTask = getContext().getRepairingTask(getServiceName());
+                repairingTask.deregisterHandler(name);
+            }
         }
 
         @Override
@@ -605,11 +603,6 @@ public class NearCachedClientMapProxy<K, V> extends ClientMapProxy<K, V> {
         public void handle(Collection<Data> keys, Collection<String> sourceUuids,
                            Collection<UUID> partitionUuids, Collection<Long> sequences) {
             repairingHandler.handle(keys, sourceUuids, partitionUuids, sequences);
-        }
-
-        private RepairingTask getRepairingTask() {
-            ClientContext clientContext = getClientContext();
-            return clientContext.getRepairingTask(SERVICE_NAME);
         }
     }
 
@@ -656,23 +649,6 @@ public class NearCachedClientMapProxy<K, V> extends ClientMapProxy<K, V> {
     // used in tests.
     public ClientContext getClientContext() {
         return getContext();
-    }
-
-    private int getConnectedServerVersion() {
-        ClientContext clientContext = getClientContext();
-        ClientClusterService clusterService = clientContext.getClusterService();
-        Address ownerConnectionAddress = clusterService.getOwnerConnectionAddress();
-
-        HazelcastClientInstanceImpl client = getClient();
-        ClientConnectionManager connectionManager = client.getConnectionManager();
-        Connection connection = connectionManager.getConnection(ownerConnectionAddress);
-        if (connection == null) {
-            logger.warning(format("No owner connection is available, "
-                    + "near cached cache %s will be started in legacy mode", name));
-            return UNKNOWN_HAZELCAST_VERSION;
-        }
-
-        return ((ClientConnection) connection).getConnectedServerVersion();
     }
 
     private boolean supportsRepairableNearCache() {

--- a/hazelcast-client/src/main/java/com/hazelcast/client/spi/ClientProxy.java
+++ b/hazelcast-client/src/main/java/com/hazelcast/client/spi/ClientProxy.java
@@ -16,6 +16,8 @@
 
 package com.hazelcast.client.spi;
 
+import com.hazelcast.client.connection.ClientConnectionManager;
+import com.hazelcast.client.connection.nio.ClientConnection;
 import com.hazelcast.client.impl.HazelcastClientInstanceImpl;
 import com.hazelcast.client.impl.protocol.ClientMessage;
 import com.hazelcast.client.impl.protocol.codec.ClientDestroyProxyCodec;
@@ -28,11 +30,14 @@ import com.hazelcast.partition.strategy.StringPartitioningStrategy;
 import com.hazelcast.spi.serialization.SerializationService;
 import com.hazelcast.util.ExceptionUtil;
 
+import java.util.Collection;
 import java.util.concurrent.Future;
+
+import static com.hazelcast.instance.BuildInfo.UNKNOWN_HAZELCAST_VERSION;
 
 /**
  * Base class for client proxies.
- *
+ * <p>
  * Allows the client to proxy operations through member nodes.
  */
 public abstract class ClientProxy implements DistributedObject {
@@ -64,6 +69,17 @@ public abstract class ClientProxy implements DistributedObject {
 
     protected final HazelcastClientInstanceImpl getClient() {
         return (HazelcastClientInstanceImpl) context.getHazelcastInstance();
+    }
+
+    protected final int getConnectedServerVersion() {
+        HazelcastClientInstanceImpl client = getClient();
+        ClientConnectionManager connectionManager = client.getConnectionManager();
+        Collection<ClientConnection> activeConnections = connectionManager.getActiveConnections();
+        for (ClientConnection connection : activeConnections) {
+            return connection.getConnectedServerVersion();
+        }
+
+        return UNKNOWN_HAZELCAST_VERSION;
     }
 
     @Deprecated

--- a/hazelcast/src/main/java/com/hazelcast/internal/nearcache/impl/invalidation/RepairingTask.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/nearcache/impl/invalidation/RepairingTask.java
@@ -22,6 +22,7 @@ import com.hazelcast.logging.ILogger;
 import com.hazelcast.spi.TaskScheduler;
 import com.hazelcast.spi.properties.HazelcastProperties;
 import com.hazelcast.spi.properties.HazelcastProperty;
+import com.hazelcast.util.ConstructorFunction;
 
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
@@ -29,6 +30,7 @@ import java.util.concurrent.RejectedExecutionException;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
 
+import static com.hazelcast.util.ConcurrencyUtil.getOrPutIfAbsent;
 import static com.hazelcast.util.Preconditions.checkNotNegative;
 import static java.lang.String.format;
 import static java.lang.System.nanoTime;
@@ -155,17 +157,28 @@ public final class RepairingTask implements Runnable {
         }
     }
 
-    public <K, V> RepairingHandler registerAndGetHandler(String name, NearCache<K, V> nearCache) {
-        RepairingHandler handler = handlers.get(name);
-        if (handler == null) {
-            handler = new RepairingHandler(name, nearCache, partitionService, localUuid, logger);
+    private class HandlerConstructor<K, V> implements ConstructorFunction<String, RepairingHandler> {
+
+        private final NearCache<K, V> nearCache;
+
+        public HandlerConstructor(NearCache nearCache) {
+            this.nearCache = nearCache;
+        }
+
+        @Override
+        public RepairingHandler createNew(String dataStructureName) {
+            RepairingHandler handler = new RepairingHandler(dataStructureName, nearCache, partitionService, localUuid, logger);
             StaleReadDetector staleReadDetector = new StaleReadDetectorImpl(handler, partitionService);
             nearCache.unwrap(DefaultNearCache.class).getNearCacheRecordStore().setStaleReadDetector(staleReadDetector);
 
             initRepairingHandler(handler);
 
-            handlers.put(name, handler);
+            return handler;
         }
+    }
+
+    public <K, V> RepairingHandler registerAndGetHandler(String dataStructureName, NearCache<K, V> nearCache) {
+        RepairingHandler handler = getOrPutIfAbsent(handlers, dataStructureName, new HandlerConstructor(nearCache));
 
         if (running.compareAndSet(false, true)) {
             scheduleNextRun();
@@ -175,8 +188,8 @@ public final class RepairingTask implements Runnable {
         return handler;
     }
 
-    public void deregisterHandler(String mapName) {
-        handlers.remove(mapName);
+    public void deregisterHandler(String dataStructureName) {
+        handlers.remove(dataStructureName);
     }
 
     /**

